Suspense tests randomly fail

**Describe the bug**

This happens for me about half the time:
```sh
 FAIL  web/test/suspense.spec.tsx
  ● Console

    console.error
      Error: Uncaught [Error: expect(received).toBe(expected) // Object.is equality
      
      Expected: true
      Received: false]
          at reportException (C:\Users\Otonashi\solid\node_modules\jsdom\lib\jsdom\living\helpers\runtime-script-errors.js:66:24)
          at Timeout.task [as _onTimeout] (C:\Users\Otonashi\solid\node_modules\jsdom\lib\jsdom\browser\Window.js:521:9)
          at listOnTimeout (node:internal/timers:557:17)
          at processTimers (node:internal/timers:500:7) JestAssertionError: expect(received).toBe(expected) // Object.is equality
      
      Expected: true
      Received: false
          at C:\Users\Otonashi\solid\packages\solid\web\test\suspense.spec.tsx:51:25
          at Timeout.task [as _onTimeout] (C:\Users\Otonashi\solid\node_modules\jsdom\lib\jsdom\browser\Window.js:516:19)
          at listOnTimeout (node:internal/timers:557:17)
          at processTimers (node:internal/timers:500:7) {
        matcherResult: {
          actual: false,
          expected: true,
          message: [Function (anonymous)],
          name: 'toBe',
          pass: false
        }
      }

      at VirtualConsole.<anonymous> (../../node_modules/jsdom/lib/jsdom/virtual-console.js:29:45)
      at reportException (../../node_modules/jsdom/lib/jsdom/living/helpers/runtime-script-errors.js:70:28)
      at Timeout.task [as _onTimeout] (../../node_modules/jsdom/lib/jsdom/browser/Window.js:521:9)

    console.error
      Error: Uncaught [Error: expect(received).toBe(expected) // Object.is equality
      
      Expected: "Hi, Jo.Hello Jo"
      Received: ""]
          at reportException (C:\Users\Otonashi\solid\node_modules\jsdom\lib\jsdom\living\helpers\runtime-script-errors.js:66:24)
          at Timeout.task [as _onTimeout] (C:\Users\Otonashi\solid\node_modules\jsdom\lib\jsdom\browser\Window.js:521:9)
          at listOnTimeout (node:internal/timers:557:17)
          at processTimers (node:internal/timers:500:7) JestAssertionError: expect(received).toBe(expected) // Object.is equality
      
      Expected: "Hi, Jo.Hello Jo"
      Received: ""
          at C:\Users\Otonashi\solid\packages\solid\web\test\suspense.spec.tsx:54:29
          at Timeout.task [as _onTimeout] (C:\Users\Otonashi\solid\node_modules\jsdom\lib\jsdom\browser\Window.js:516:19)
          at listOnTimeout (node:internal/timers:557:17)
          at processTimers (node:internal/timers:500:7) {
        matcherResult: {
          actual: '',
          expected: 'Hi, Jo.Hello Jo',
          message: [Function (anonymous)],
          name: 'toBe',
          pass: false
        }
      }

      at VirtualConsole.<anonymous> (../../node_modules/jsdom/lib/jsdom/virtual-console.js:29:45)
      at reportException (../../node_modules/jsdom/lib/jsdom/living/helpers/runtime-script-errors.js:70:28)
      at Timeout.task [as _onTimeout] (../../node_modules/jsdom/lib/jsdom/browser/Window.js:521:9)

  ● Testing Suspense › Toggle with refresh transition

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

      49 |     setTimeout(() => {
      50 |       expect(div.innerHTML).toBe("Hi, .Hello ");
    > 51 |       expect(pending()).toBe(true);
         |                         ^
      52 |     });
      53 |     setTimeout(() => {
      54 |       expect(div.innerHTML).toBe("Hi, Jo.Hello Jo");

      at web/test/suspense.spec.tsx:51:25
      at Timeout.task [as _onTimeout] (../../node_modules/jsdom/lib/jsdom/browser/Window.js:516:19)

  ● SuspenseList › revealOrder forwards

    expect(received).toBe(expected) // Object.is equality

    Expected: "Hi, Jo.Hello Jo"
    Received: ""

      52 |     });
      53 |     setTimeout(() => {
    > 54 |       expect(div.innerHTML).toBe("Hi, Jo.Hello Jo");
         |                             ^
      55 |       expect(pending()).toBe(false);
      56 |       expect(finished).toBe(true);
      57 |       done();

      at web/test/suspense.spec.tsx:54:29
      at Timeout.task [as _onTimeout] (../../node_modules/jsdom/lib/jsdom/browser/Window.js:516:19)

Summary of all failing tests
 FAIL  web/test/suspense.spec.tsx
  ● Testing Suspense › Toggle with refresh transition

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

      49 |     setTimeout(() => {
      50 |       expect(div.innerHTML).toBe("Hi, .Hello ");
    > 51 |       expect(pending()).toBe(true);
         |                         ^
      52 |     });
      53 |     setTimeout(() => {
      54 |       expect(div.innerHTML).toBe("Hi, Jo.Hello Jo");

      at web/test/suspense.spec.tsx:51:25
      at Timeout.task [as _onTimeout] (../../node_modules/jsdom/lib/jsdom/browser/Window.js:516:19)

  ● SuspenseList › revealOrder forwards

    expect(received).toBe(expected) // Object.is equality

    Expected: "Hi, Jo.Hello Jo"
    Received: ""

      52 |     });
      53 |     setTimeout(() => {
    > 54 |       expect(div.innerHTML).toBe("Hi, Jo.Hello Jo");
         |                             ^
      55 |       expect(pending()).toBe(false);
      56 |       expect(finished).toBe(true);
      57 |       done();

      at web/test/suspense.spec.tsx:54:29
      at Timeout.task [as _onTimeout] (../../node_modules/jsdom/lib/jsdom/browser/Window.js:516:19)

Test Suites: 1 failed, 20 passed, 21 total
Tests:       2 failed, 316 passed, 318 total
Snapshots:   0 total
Time:        6.602 s
Ran all test suites.
```

**To Reproduce**
1. Run `npm test`
2. Maybe the test fails?
3. Rerun `npm test` until the test doesn't fail

**Expected behavior**
The behaviour should be deterministic. 

**Additional context**
Windows 10 Pro 21H1 Build 19043.1348
npm 8.1.1

--- a/packages/solid/web/test/suspense.spec.tsx
+++ b/packages/solid/web/test/suspense.spec.tsx
@@ -3,10 +3,18 @@ import "../../test/MessageChannel";
 import { lazy, createSignal, createResource, useTransition, enableScheduling } from "../../src";
 import { render, Suspense, SuspenseList } from "../src";
 
+
 global.queueMicrotask = setImmediate;
 enableScheduling();
 
+beforeEach(() => {
+  jest.useFakeTimers();
+})
+afterEach(() => {
+  jest.useRealTimers();
+})
 describe("Testing Suspense", () => {
+
   let div = document.createElement("div"),
     disposer: () => void,
     resolvers: Function[] = [],
@@ -32,31 +40,44 @@ describe("Testing Suspense", () => {
     expect(div.innerHTML).toBe("Loading");
   });
 
-  test("Toggle Suspense control flow", done => {
+  test("Toggle Suspense control flow", async (done) => {
     for (const r of resolvers) r({ default: ChildComponent });
-    setTimeout(() => {
+
+    queueMicrotask(() => {
       expect(div.innerHTML).toBe("Hi, .Hello ");
       done();
     });
   });
 
-  test("Toggle with refresh transition", done => {
+  test("Toggle with refresh transition", async (done) => {
     const [pending, start] = useTransition();
     let finished = false;
+
     start(() => trigger(true)).then(() => (finished = true));
     expect(div.innerHTML).toBe("Hi, .Hello ");
     expect(finished).toBe(false);
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("Hi, .Hello ");
-      expect(pending()).toBe(true);
-      expect(finished).toBe(false);
-    });
-    setTimeout(() => {
+    // wait trigger resource refetch
+    await Promise.resolve();
+
+    expect(div.innerHTML).toBe("Hi, .Hello ");
+    expect(pending()).toBe(true);
+    expect(finished).toBe(false);
+    
+    // Exhausts create-resource setTimeout
+    jest.runAllTimers();
+    // wait update suspence state
+    await Promise.resolve();
+    // wait update computation 
+    jest.runAllTicks();
+    jest.runAllTimers();
+    // wait write signal succ
+    queueMicrotask(()=>{
       expect(div.innerHTML).toBe("Hi, Jo.Hello Jo");
       expect(pending()).toBe(false);
       expect(finished).toBe(true);
-      done();
-    }, 400);
+      done()
+    });
+    jest.runAllTicks();
   });
 
   test("dispose", () => {
@@ -87,7 +108,7 @@ describe("SuspenseList", () => {
       return <div>{value()}</div>;
     };
 
-  test("revealOrder together", done => {
+  test("revealOrder together", async () => {
     const div = document.createElement("div"),
       Comp = () => (
         <SuspenseList revealOrder="together">
@@ -104,20 +125,20 @@ describe("SuspenseList", () => {
       );
     const dispose = render(Comp, div);
     expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
-    }, 110);
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
-    }, 210);
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>C</div>");
-      dispose();
-      done();
-    }, 310);
+    jest.advanceTimersByTime(110);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
+    jest.advanceTimersByTime(100);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
+    jest.advanceTimersByTime(100);
+    // wait effect update
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>C</div>");
+    dispose();
   });
 
-  test("revealOrder forwards", done => {
+  test("revealOrder forwards", async () => {
     const div = document.createElement("div"),
       Comp = () => (
         <SuspenseList revealOrder="forwards">
@@ -134,20 +155,22 @@ describe("SuspenseList", () => {
       );
     const dispose = render(Comp, div);
     expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
-    }, 110);
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>Loading 3</div>");
-    }, 210);
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>C</div>");
-      dispose();
-      done();
-    }, 310);
+
+    jest.advanceTimersByTime(110);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
+    
+    jest.advanceTimersByTime(100);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>Loading 3</div>");
+    
+    jest.advanceTimersByTime(100);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>C</div>");
+    dispose();
   });
 
-  test("revealOrder forwards hidden", done => {
+  test("revealOrder forwards hidden", async () => {
     const div = document.createElement("div"),
       Comp = () => (
         <SuspenseList revealOrder="forwards" tail="hidden">
@@ -164,20 +187,23 @@ describe("SuspenseList", () => {
       );
     const dispose = render(Comp, div);
     expect(div.innerHTML).toBe("");
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("");
-    }, 110);
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>A</div><div>B</div>");
-    }, 210);
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>C</div>");
-      dispose();
-      done();
-    }, 310);
+
+    jest.advanceTimersByTime(110);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("");
+    
+    jest.advanceTimersByTime(100);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>A</div><div>B</div>");
+    
+    
+    jest.advanceTimersByTime(100);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>C</div>");
+    dispose();
   });
 
-  test("revealOrder forwards", done => {
+  test("revealOrder forwards", async () => {
     const div = document.createElement("div"),
       Comp = () => (
         <SuspenseList revealOrder="forwards">
@@ -194,20 +220,22 @@ describe("SuspenseList", () => {
       );
     const dispose = render(Comp, div);
     expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
-    }, 110);
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>Loading 3</div>");
-    }, 210);
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>C</div>");
-      dispose();
-      done();
-    }, 310);
+
+    jest.advanceTimersByTime(110);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
+
+    jest.advanceTimersByTime(100);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>Loading 3</div>");
+
+    jest.advanceTimersByTime(100);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>C</div>");
+    dispose();
   });
 
-  test("revealOrder forwards collapse", done => {
+  test("revealOrder forwards collapse", async () => {
     const div = document.createElement("div"),
       Comp = () => (
         <SuspenseList revealOrder="forwards" tail="collapsed">
@@ -224,20 +252,22 @@ describe("SuspenseList", () => {
       );
     const dispose = render(Comp, div);
     expect(div.innerHTML).toBe("<div>Loading 1</div>");
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>Loading 1</div>");
-    }, 110);
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>Loading 3</div>");
-    }, 210);
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>C</div>");
-      dispose();
-      done();
-    }, 310);
+
+    jest.advanceTimersByTime(110);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>Loading 1</div>");
+
+    jest.advanceTimersByTime(100);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>Loading 3</div>");
+
+    jest.advanceTimersByTime(100);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>C</div>");
+    dispose();
   });
 
-  test("revealOrder backwards collapse", done => {
+  test("revealOrder backwards collapse", async () => {
     const div = document.createElement("div"),
       Comp = () => (
         <SuspenseList revealOrder="backwards" tail="collapsed">
@@ -254,20 +284,22 @@ describe("SuspenseList", () => {
       );
     const dispose = render(Comp, div);
     expect(div.innerHTML).toBe("<div>Loading 3</div>");
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>Loading 3</div>");
-    }, 110);
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>Loading 3</div>");
-    }, 210);
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>C</div>");
-      dispose();
-      done();
-    }, 310);
+
+    jest.advanceTimersByTime(110);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>Loading 3</div>");
+
+    jest.advanceTimersByTime(100);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>Loading 3</div>");
+
+    jest.advanceTimersByTime(100);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>C</div>");
+    dispose();
   });
 
-  test("nested SuspenseList together", done => {
+  test("nested SuspenseList together", async () => {
     const div = document.createElement("div"),
       Comp = () => (
         <SuspenseList revealOrder="together">
@@ -288,20 +320,22 @@ describe("SuspenseList", () => {
       );
     const dispose = render(Comp, div);
     expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
-    }, 110);
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
-    }, 210);
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>C</div>");
-      dispose();
-      done();
-    }, 310);
+
+    jest.advanceTimersByTime(110);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
+
+    jest.advanceTimersByTime(100);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
+
+    jest.advanceTimersByTime(100);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>C</div>");
+    dispose();
   });
 
-  test("nested SuspenseList forwards", done => {
+  test("nested SuspenseList forwards", async () => {
     const div = document.createElement("div"),
       Comp = () => (
         <SuspenseList revealOrder="forwards">
@@ -322,16 +356,18 @@ describe("SuspenseList", () => {
       );
     const dispose = render(Comp, div);
     expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
-    }, 110);
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>Loading 3</div>");
-    }, 210);
-    setTimeout(() => {
-      expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>C</div>");
-      dispose();
-      done();
-    }, 310);
+
+    jest.advanceTimersByTime(110);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>Loading 1</div><div>Loading 2</div><div>Loading 3</div>");
+
+    jest.advanceTimersByTime(100);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>Loading 3</div>");
+
+    jest.advanceTimersByTime(100);
+    await Promise.resolve();
+    expect(div.innerHTML).toBe("<div>A</div><div>B</div><div>C</div>");
+    dispose();
   });
 });

